### ISSUE TITLE
proposal `$switch`

### ISSUE THREAD

**Issue by jonasfj** [taskcluster/json-e#161]
```yaml
---
title: $switch on string
context: {key: "kindA"}
template:
  $switch: "key"
  kindA: {$eval: '4 - 1'}
  kindB: {$eval: '3 + 4'}
result: 3
---
title: $switch on type
context: {key: 42}
template:
  $switch: "type(key)"  # works when bstacks type/typeof function is introduced
  string: {$eval: '4 - 1'}
  number: {$eval: '3 + 4'}
result: 7
---
title: $switch on number
context: {key: 42}
template:
  $switch: "key"
  kindA: {$eval: '4 - 1'}
  "42": {$eval: '3 + 4'}
result: 7
---
title: $switch without match
context: {key: 42}
template:
  $switch: "key"
  "43": 1
result: null
---
title: $switch without branches
context: {key: 42}
template:
  $switch: "key"
result: null
---
title: $switch without match as deleteMarker
context: {key: 42}
template:
  required: 23
  optional:
    $switch: "key"
    "43": 1
result: {required: 23}
---
```

**Comment by djmitche** [comment]
What's the use-case for this?  I'm not opposed except on the basis of "avoid unnecessary complexity".

**Comment by jonasfj** [comment]
It's thought as a general alternative to if-elseif which we don't have.
But in particular I think it would be useful with a `typeof/type` function to allow type switching:
```yaml
$switch: "type(key)"  # works when bstacks type/typeof function is introduced
  string: 7
  number: {$eval: '3 + key'}
  object: {$eval: 'key.value'}
```

**Comment by jonasfj** [comment]
Imagine updating artifacts, you could switch on `artifacts[i].kind` and inside a `$map` to treat folder and file artifacts differently.

**Comment by djmitche** [comment]
OK, I'm sold :)

**Comment by djmitche** [comment]
I'm writing `.taskcluster.yml` for Gecko, and it would be useful there, too.

**Comment by djmitche** [comment]
@owlishdeveloper one of the things I don't like about the proposal here is that it stringifies the value:

```yaml
$switch: somevalue
13: "thirteen"
twelve: "twelve"
```

Then both `13` (the number) and `"13"` (the string) will produce `"thirteen"` as output.

You suggested `$match`, and that could potentially contain condition expressions in strings.  Something like
```yaml
$match:
  - "somevalue == 13": "thirteen (number)"
  - "somevalue == '13'": "thirteen (string)"
  - "somevalue == 'twelve'": "twelve"
```

This is much closer to an if/elseif structure in that it's a series of conditionals, instead of a lookup of an object property.  But it's more flexible, too (`"somevalue > 10 && somevalue < 20"`).  Note that I used a YAML list here so that the conditions are evaluated in the order they are written.

**Comment by owlishDeveloper** [comment]
Yes, that's exactly the idea, only I suggested `$when` because that way you can do not only:

```yml
$when:
  - "ref == 'master' && event == push": blablabla
  - "ref == 'master' || event == release": blablabla
```
But also check type:

```yml
$when: "x":
  - $is "string": blablabla
  - $is "number": blablabla
```
or the presence of a value in some array that comes from context:

```yml
$when: "branch":
  - $in "event.branches": blablabla
  - $else: blablabla
```
(The above basically just shamelessly stolen from kotlin)

**Comment by owlishDeveloper** [comment]
(I'm not insistent on the naming, the important thing is what it can do)

**Comment by djmitche** [comment]
We still need to stick to YAML formatting, so everything before the `:` is a single string:

```yaml
$when: "x"
"$is 'string'": blahblah
```

equivalent to JSON `{"$when": "x", "$is 'string'": "blahblah"}`.

I'm also concerned about adding a bunch of new syntax like `$is` and `$in`.  In the syntax I suggested earlier, the object keys are just regular JSON-e expressions (like those used for `$if`).  The more Kotlin-style "when (<something>) { <match-expr> -> <value>, <match-expr> -> <value> }" syntax, then we need to invent match expressions -- and people expect a lot of things from match expressions!

I think the simpler thing to implement, and the simpler thing for our users, is conditional expressions.

**Comment by owlishDeveloper** [comment]
`$is` and `$in` are not needed for our purposes of simplifying the interface, it's just something that can be considered down the road if we need it.

Anyways, I do like the idea you are suggesting with `$match` there - it looks nice and simple.

**Comment by owlishDeveloper** [comment]
The only thing I'd like to note is that it would be lovely if it weren't quite `switch` or `when` in the sense that they will let you go down just single branch (the one that matches). I would love this thingy to allow us to match _everything_ that matches. so that we could have

```yml
$match:
  - "event == push": task1
  - "event == push && branch == master": task 2
  - "event == pull_request": task3
```
yielding Tasks[]: [task1, task2] for a push to master and [tas1] for a push to some other branch.

**Comment by djmitche** [comment]
I think that would surprise people who are used to `match` (in other languages) picking exactly one alternative.  It also means that in the case where I want just one alternative, I'd get a list instead:

```yaml
$match:
  - "x == 10": "ten"
  - "x == 20": "twenty"
```
would yield `[]` or `["ten"]` or `["twenty"]`, which is pretty surprising.

We can already do what you're describing with `$if` without any awful indentation or nesting:

```yaml
- $if: 'event == "push"'
  then: task1
- $if: 'event == "push" && branch == "master"'
  then: task2
- $if: 'event == "pull_request"'
  then: task3
```

**Comment by djmitche** [comment]
I think I have gotten lost from the case we're trying to solve.  Can you link to an example where things are awkward, and $switch or $match might help?

**Comment by owlishDeveloper** [comment]
https://github.com/taskcluster/taskcluster-cli/blob/master/.taskcluster.yml

**Comment by djmitche** [comment]
OK, that's

```yaml
  - $if: 'tasks_for == "github-pull-request" && event["action"] in ["opened", "reopened", "synchronize"]'
    then:
      $if: 'event.pull_request.base.ref == "master"'
      then: ...
    $else:
      $if: 'tasks_for == "github-push"'
      then:
        $if: 'event.ref == "refs/heads/master"'
        then: ...
```

Is the main problem that we can't combine those into `(tasks_for == "github-pull-request" && event["action"] in ["opened", "reopened", "synchronize"] && event.pull_request.base.ref == "master") || (tasks_for == "github-push" && event.ref == "refs/heads/master")`?  I think that's actually due to #244 -- we can't refer to `event.pull_request` in that expression because the field might not exist.  But in most programming languages, once `tasks_for == "github-pull-request"` was evaluated and found to be false, the part after `&&` wouldn't even be evaluated, so there'd be no error.

**Comment by owlishDeveloper** [comment]
It would be lovely to address #244 as well. That was something of a surprise/annoyance to me too when I was working on `cli`'s yaml

**Comment by djmitche** [comment]
Even if it's not the exact solution to this issue, I think $switch / $match is going to be a good addition to the language.. so maybe we should get implementing? :)

**Comment by owlishDeveloper** [comment]
yaaaaassss 😋

**Comment by djmitche** [comment]
Fixed in #251.
